### DISCUSSION TITLE
Does a shadow variable update when a source variable is set to null?

### DISCUSSION THREAD

**Discussion by mikepictor** [TimefoldAI/timefold-solver#856] [General]
I have a shadow variable watching 2 properties. One in the same class that it is declared in, one watching a nullable planning variable in a different class.

The shadow variable is meant to balance both planning variables and compute an end time of a task. I have noticed some examples where my end time is shorter than it should be based on the end state of the planning variables. The first variable always has a value, but the second, being nullable, may not. I am just wondering whether the shadow variable may get computed when the 2nd property gets a new value, but maybe is NOT computed if the second property is assigned a null value?

I am on Timefold 1.6.0

**Comment by mikepictor** [comment]
to be clear, the end time is never to long, it's usually entirely correct, and occasionally..it's too short. The variable being null in my case would create a longer interval, which is what made me suspicious of that scenario.

**Comment by triceo** [comment]
Hello @mikepictor! There is no code in the solver which would skip shadow variable updates under any circumstances; a successful triggering of shadow variables leaves the solution in a consistent state, with all shadow variables on all entities having been updated.

However, the solver can only recompute what it knows to have been changed. Are you sure your variable listeners/custom moves are correctly calling all variable changes? If so, this may be a bug, in which case I will ask you to please submit a simplified reproducible code that I can run to see the issue on my end.

**Reply by mikepictor** [reply]
I can't reliably reproduce it in my own code, only the observation that sometimes (usually not, but sometimes) my tasks are too short, and the pure guess that maybe it's because it didn't detect the property being set to null. The only way this can happen is for the shadow variable to compute when the value was assigned, and then for the value to remain when the value gets unassigned again, creating an incorrect result

I just upgraded to 1.10, and I see that `nullable` has been replaced by `allowUnassigned`, so I am going to update that at least, and monitor to see if I can spot the problem ever re-ocurring. 

I can't produce a reproducable scenario, as I don't really know how to reproduce, but being 4 versions behind probably doesn't help.
